### PR TITLE
feat(python): merge_insert in async Python

### LINKED ISSUE #1401: bug(python): async table merge_insert cannot be waited for
### LanceDB version

v0.8.2

### What happened?

 async table merge_insert function produces the following warning

```python
import pyarrow as pa
import lancedb

class MergeList:
    def __init__(self, items):
        self.items = items

    def iter(self):
        return iter(self.items)

data = pa.table({ 
                  "c1": ["A", "B"], 
                  "c2": ["a", "b"],
                  "c3": [1, 2]
               })
db = await lancedb.connect_async("lancedb")
await db.create_table("my_table", data)
table = await db.open_table("my_table")
new_data = pa.table({ 
                    "c1": ["A", "B",], 
                    "c2": ["a", "b"], 
                    "c3": [1,4]
               })

merge_cols = MergeList(["c1", "c2"])
table.merge_insert(merge_cols).when_matched_update_all().when_not_matched_insert_all().execute(new_data)

```

/usr/local/lib/python3.10/dist-packages/lancedb/merge.py:107: RuntimeWarning: coroutine 'AsyncTable._do_merge' was never awaited
  self._table._do_merge(self, new_data, on_bad_vectors, fill_value)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

---

Trying to await on table.merge_insert , produces the following error

```python 
await table.merge_insert(merge_cols).when_matched_update_all().when_not_matched_insert_all().execute(new_data)

```
await table.merge_insert(merge_cols).when_matched_update_all().when_not_matched_insert_all().execute(new_data)

TypeError: object NoneType can't be used in 'await' expression

### Are there known steps to reproduce?

_No response_

### PR DESCRIPTION
**Pull request by wjones127**
Fixes #1401

### CODE REVIEW

File: python/python/lancedb/table.py [hunk 1]
```diff
@@ -2464,7 +2464,31 @@ async def _do_merge(
         on_bad_vectors: str,
         fill_value: float,
     ):
-        pass
```
Comment (westonpace) [inline python/python/lancedb/table.py:2467]: Huh...was this just a TODO then?

File: python/python/lancedb/table.py [hunk 2]
```diff
@@ -2464,7 +2464,31 @@ async def _do_merge(
...
+        if fill_value is None:
+            fill_value = 0.0
+        data, _ = _sanitize_data(
+            new_data,
+            schema,
+            metadata=schema.metadata,
+            on_bad_vectors=on_bad_vectors,
+            fill_value=fill_value,
+        )
+        if isinstance(data, pa.Table):
```
Comment (westonpace) [inline python/python/lancedb/table.py:2479]: Does `_sanitize_data` ever return something that isn't `pa.Table`?

File: python/python/lancedb/table.py [hunk 3]
```diff
@@ -2464,7 +2464,31 @@ async def _do_merge(
...
+            data = pa.RecordBatchReader.from_batches(data.schema, data.to_batches())
+        return await self._inner.execute_merge_insert(
```
Comment (westonpace) [inline python/python/lancedb/table.py:2481]: What's being returned here?

**Review by westonpace — APPROVED** [review]

File: python/python/lancedb/table.py (hunk 2 above)
Comment (wjones127) [reply python/python/lancedb/table.py:2479]: I'm not sure. This is taken from `AsyncTable.add()`. I think right now it's always a table. It should really be better typed...

File: python/python/lancedb/table.py (hunk 3 above)
Comment (wjones127) [reply python/python/lancedb/table.py:2481]: Nothing. I can take away the return.

File: python/python/lancedb/table.py (hunk 1 above)
Comment (wjones127) [reply python/python/lancedb/table.py:2467]: Yeah was left as TODO.

**Pull request merged**